### PR TITLE
Automatically fix .gitreview rebase errors

### PR DESCRIPTION
**Pull request by javierpena**
We are often finding an issue when rebasing, where the upstream branch
is changed and a downstrem .gitreview file change fails to rebase.
This commit fixes this by:

- Attempting to detect .gitreview related rebases
- If that happens, skip the .gitreview patch and re-create it at
  the end of the rebase.

### CODE REVIEW

File: patch_rebaser/patch_rebaser.py [hunk 1]
```diff
@@ -228,10 +228,18 @@ def perform_rebase(self):
...
-        except git_exceptions.RebaseException:
-            LOGGER.info("Could not rebase. Cleaning up.")
-            self.repo.branch.abort_rebase()
-            raise
+        except git_exceptions.RebaseException as e:
+            if '.gitreview' in str(e):
+                LOGGER.warning("A patch including the .gitreview file failed "
+                               "to rebase, skipping it.")
+                self.repo.git.rebase('--skip')
+                self.perform_rebase()   # Just in case, rebase again
```
Comment (fuzzball81) [inline patch_rebaser/patch_rebaser.py]: Do we want to wrap this in a try...except as well just to catch any other rebase issues? We may need to consider having a loop of; rebase, auto-mitigate, continue until we either hit an issue where we can't auto-mitigate or we complete the rebase. That would require more changes, and would be out of the scope of this PR, but something we could do as a next step.

File: patch_rebaser/patch_rebaser.py [hunk 2]
```diff
@@ -261,6 +269,28 @@ def update_remote_patches_branch(self):
...
 
+    def _rebuild_gitreview(self):
+        dlrn = get_dlrn_variables()
+        url = self.repo.repo.remote(self.remote).url
+        if isinstance(url, list):
+            url = url[0]
+        component = url.split('/')[-1]
+        host_port = url.split('/')[2]
+        host = host_port.split(':')[0]
+        port = host_port.split(':')[1]
```
Comment (fuzzball81) [inline patch_rebaser/patch_rebaser.py]: Let's split the url into a list once and do the same for the host_port once we get it. I would also add a comment describing the expected format of the strings so it's clear what we are trying to extract.

File: patch_rebaser/patch_rebaser.py (hunk 1 above)
Comment (jguiditta) [reply patch_rebaser/patch_rebaser.py]: Agree, I think as a next step this process needs to be decomposed.  Actually, I was just thinking, maybe we split out the error handling to a separate method as part of that.  This code feels to me like the beginnings of a decision tree, and as soon as we consider adding one more if here, we should reassess and make such a mechanism, imo.

File: patch_rebaser/patch_rebaser.py [hunk 3]
```diff
@@ -228,10 +228,18 @@ def perform_rebase(self):
...
-            self.repo.branch.abort_rebase()
-            raise
+        except git_exceptions.RebaseException as e:
+            if '.gitreview' in str(e):
+                LOGGER.warning("A patch including the .gitreview file failed "
+                               "to rebase, skipping it.")
+                self.repo.git.rebase('--skip')
+                self.perform_rebase()   # Just in case, rebase again
+                LOGGER.info("Re-creating .gitreview change")
+                self._rebuild_gitreview()
```
Comment (jguiditta) [inline patch_rebaser/patch_rebaser.py]: Actually, just had another thought that would make this chunk feel (to me) a bit less intertwined.  What if we move out the git.rebase and perform_rebase calls out of this exception, and into _rebuild_gitreview?  then it feels less like we are doing extra work within the exception itself, even though perhaps technically it still is.  Another option would be to extend RebaseException to handle this code (which could be done a couple ways if we went that path).  However, if everyone else is ok with this, I am willing to let it be as is for now and consider other implementation later.

**Comment by javierpena** [comment]
Thanks everyone for the comments. I have pushed a new version of the commit, including unit tests, taking some reusable code to functions, and creating the loop suggested by @fuzzball81. Let me know how it looks now.

File: tests/test_patch_rebaser.py [hunk 4]
```diff
@@ -322,3 +323,59 @@ def test_packages_to_process_skips_packages_not_in_the_list(
...
+def test_try_automated_rebase_fix(mock_repo):
+    rebaser = Rebaser(mock_repo, "my_branch", "my_commit", "my_remote",
+                      "my_tstamp", dev_mode=True)
+    rebaser._rebuild_gitreview = Mock()
+    exception = exceptions.RebaseException('.gitreview failed to rebase')
+    output = rebaser.try_automated_rebase_fix(exception)
+
+    mock_repo.git.rebase.assert_called()
+    rebaser._rebuild_gitreview.assert_called()
+    assert output == True
```
Comment (jguiditta) [inline tests/test_patch_rebaser.py]: Minor error in the CI on this, wants it to be:
```
assert output is True
```

File: tests/test_patch_rebaser.py [hunk 5]
```diff
@@ -322,3 +323,59 @@ def test_packages_to_process_skips_packages_not_in_the_list(
         main()
 
     mock_gitrepo.assert_not_called()
+
+
+def test_rebase_exception_gitreview(mock_repo):
```
Comment (jguiditta) [inline tests/test_patch_rebaser.py]: Thank you for adding these tests!  If you don't mind, one more minor request on them.  If you look at the others that @jpichon has added we have a little docstring block at the top which explains the expected behavior under test (in 'Given, When, Then' format).  If you could, please add this, it makes it much easier for others to come in and work with the code, as they can more quickly understand what it is attempting to do.

File: tests/test_patch_rebaser.py (hunk 4 above)
Comment (javierpena) [reply tests/test_patch_rebaser.py]: Oh, I didn't know it was required, actually it passed tests ocally without a complain :?. Anyway, it's fixed in the latest commit.

File: tests/test_patch_rebaser.py (hunk 5 above)
Comment (javierpena) [reply tests/test_patch_rebaser.py]: Done, thanks for the review!

**Pull request merged**